docs/constraints/TypeScript.md

```diff
@@ -0,0 +1,38 @@
+# TypeScript Constraints
+
+These are the TypeScript specific constraints applied to the MetaModel before reaching the generators.
+
+## Model Naming
+These are the constraints that is applied to model naming. The `Rule key` is what you can use in the options to overwrite the default behavior. See [constraint customization](../constraints.md#Customization).
+
+|Rule key|Rule|Resolution|
+|---|---|---|
+|NO_SPECIAL_CHAR|No special characters| Special characters are replaced by their name, for example `!` is replaced with `exclamation`. For typescript `_` and `$` are an exception to this rule. |
+|NO_NUMBER_START_CHAR|No numbers as starting characters|Default behavior is pre pending `number_` in front of the first character|
+|NO_EMPTY_VALUE|No empty values|Default behavior is to use `empty` as name. |
+|NO_RESERVED_KEYWORDS|No reserved keywords|TypeScript has a list of reserved keywords ([see the full list here](https://github.com/asyncapi/modelina/blob/4e3604828c92c8951731e0fe569887b1162cbaf1/src/generators/typescript/Constants.ts#L2))|
+|NAMING_FORMATTER|Must be formatted equally|Model name is formatted using pascal case|
```

Comment:
I did not really think about it much. 

I mean you can see it as a boolean kind of expression, `once run, there should be no xxx`. 

What would you say makes more sense in terms of naming?

Reply:
With a smaller name and without `no_` prefix, because that means that you can switch on or off - e.g. 'emptyValue'. On the other hand, we have two choices, either we do as we currently do with so many functions, or only one simple one, but we will expose helpers that we use in default one, something like that:

```ts
function defaultPropertyKeyConstraints(propertyName, ...options) {
  let constrainedPropertyName = propertyName;
  constrainedPropertyName = NO_SPECIAL_CHAR(constrainedPropertyName);
    constrainedPropertyName = NO_NUMBER_START_CHAR(constrainedPropertyName);
    constrainedPropertyName = NO_EMPTY_VALUE(constrainedPropertyName);
    constrainedPropertyName = NO_RESERVED_KEYWORDS(constrainedPropertyName);
    constrainedPropertyName = NAMING_FORMATTER(constrainedPropertyName);
    constrainedPropertyName = NO_DUPLICATE_PROPERTIES(constrainedObjectModel, objectModel, constrainedPropertyName, constrainRules.modelPropertyKeyConstraints.NAMING_FORMATTER);
}
```

and then interface for constraints:

```ts
export interface TypeScriptConstraints {
  enumKey: ...,
  modelNameKey: ...,
  propertyKey: (propertyName, ...options) => string,
}
```

Also we can create factory for our default function for constraint and here to have a separation (in the options of this function) for the dozen or so functions that someone could overwrite, something like:

```ts
interface DefaultPropertyKeyConstraintsOptions {
  specialChar: (...) => string;
}

function defaultPropertyKeyConstraints(options: DefaultPropertyKeyConstraintsOptions) {
  return () => { ... } // in this function we will use functions passed in the DefaultPropertyKeyConstraintsOptions
}
```

In this case people will be able to create their own "preset" for constraints by creating factory like above. What do you think?

Reply:
Yep :) To not having several functions for single type like `propertyKey` or `enumKey` but single one and then people will be able to create their own `constrain function` with factory.

Reply:
I like it 👍